Chore: Provide function which runs after possible errors occured

danilminibaev opened the pull request:
Provide `onError` function, to call extra actions in case of validation errors are there (both client-side and server-side).
E.g. to be able to toggle scroll into field with error.

Commit:
Chore: Added onError function & tests

In case to invoke some additional actions (e.g. scrolling to
field with error) 'onError' function was added

ingmarh commented on src/form/Form.jsx:
```diff
@@ -166,6 +287,21 @@ export default class Form extends React.Component {
+    const firstErrorKeyPath = pathList && pathList[0] // first path list
+
+    if (firstErrorKeyPath && firstErrorKeyPath.length > 0) {
+      this.props.scrollIntoError && this.props.scrollIntoError(firstErrorKeyPath)
```
Since you have this in "defaultProps", you can omit the check.

ingmarh commented on src/form/Form.jsx:
```diff
@@ -79,8 +100,11 @@ export default class Form extends React.Component {
+    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
+    externalErrors: ImmutablePropTypes.map,
+    handleUnmappedErrors: PropTypes.func,
+    scrollIntoError: PropTypes.func
```
Instead of having a prop called "scrollIntoError", I suggest to have a more general "onError" prop (receiving the current errors as argument).

Scrolling (or other actions) could then be handled by using this prop.

Commit:
Fix: Test adjustment

Commit:
Fix: Naming fix

Commit:
Fix: Renamed 'errorHandler' to 'handleErrors'

depoulo commented on package.json:
```diff
@@ -88,5 +89,8 @@
     "cacheDirectory": "<rootDir>/.cache/jest"
   },
-  "false": {}
+  "false": {},
```
wonder what that is 😬

danilminibaev replied:
🤔

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -465,4 +466,46 @@ describe('Form', function () {
+    const onError = sinon.stub()
+
+    // server side error path
+    const otherPros = {
```
typo: otherProps

danilminibaev replied:
Will fix 👍

depoulo reviewed (changes requested):
just the typo

Commit:
Chore: Added `onError` call in every error case, removed redunant code

Decided to move redundant (for this soultion) code into `ng-merchant-ui`
(here it was searching for first error path)
and to call `onError` (passed from somewhere else) on every possible
error case

Commit:
Fix: Adjusted test due to latest changes

Commit:
Fix: Passing down explicitly `fromSubmit` flag

Commit:
Fix: Typo

depoulo committed:
Merge branch 'master' into chore/onerror-function

depoulo committed:
Update package.json

depoulo committed:
Update package-lock.json
